com.unity.netcode.gameobjects/Runtime/Components/Helpers/ComponentController.cs

```diff
@@ -0,0 +1,543 @@
+        {
+            return !(component.GetType().IsSubclassOf(typeof(NetworkBehaviour)) || component.GetType() == typeof(NetworkObject) || component.GetType() == typeof(NetworkManager));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static string GetComponentNameFormatted(Object component)
+        {
+            // Split the class name up based on capitalization
+            var classNameDisplay = Regex.Replace(component.GetType().Name, "([A-Z])", " $1", RegexOptions.Compiled).Trim();
+            return $"{component.name} ({classNameDisplay})";
+        }
+
+        /// <inheritdoc/>
+        /// <remarks>
+        /// Checks for invalid <see cref="Object"/> entries.
+        /// </remarks>
+        protected virtual void OnValidate()
+        {
+            if (Components == null || Components.Count == 0)
+            {
+                return;
+            }
+
+            var gameObjectsToScan = new List<ComponentEntry>();
+            // First pass is to verify all entries are valid and look for any GameObjects added as an entry to process next
+            for (int i = Components.Count - 1; i >= 0; i--)
+            {
+                if (Components[i] == null)
+                {
+                    continue;
+                }
+
+                if (Components[i].Component == null)
+                {
+                    continue;
+                }
+                var objectType = Components[i].Component.GetType();
+                if (objectType == typeof(GameObject))
+                {
+                    if (!gameObjectsToScan.Contains(Components[i]))
+                    {
+                        gameObjectsToScan.Add(Components[i]);
+                    }
+                    Components.RemoveAt(i);
+                    continue;
+                }
+
+                if (!IsValidComponentType(Components[i].Component))
+                {
+                    Debug.LogWarning($"Removing {GetComponentNameFormatted(Components[i].Component)} since {Components[i].Component.GetType().Name} is not an allowed component type.");
+                    Components.RemoveAt(i);
+                    continue;
+                }
+
+                var propertyInfo = Components[i].Component.GetType().GetProperty("enabled", BindingFlags.Instance | BindingFlags.Public);
+                if (propertyInfo == null && propertyInfo.PropertyType != typeof(bool))
+                {
+                    Debug.LogWarning($"{Components[i].Component.name} does not contain a public enabled property! (Removing)");
+                    Components.RemoveAt(i);
+                }
+            }
+
+            // Second pass is to process any GameObjects added.
+            // Scan the GameObject and all of its children and add all valid components to the list.
+            foreach (var entry in gameObjectsToScan)
+            {
+                var asGameObject = entry.Component as GameObject;
```

Comment:
Does it make sense to change `gameObjectsToScan` from `List<ComponentEntry>` to `List<GameObject>`?

Reply:
Thanks for the explanation! 

I'm seeing that `gameObjectsToScan` is created as `List<ComponentEntry>` ([line 226](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/3518/files/fe889232aa97af99528452eebaf177b9919e5c00#diff-c8dda3f44ea07cbcb7f349be0db86ec4812b54eeae5adc40046c73151d09082aR226)) and then the type of the component is checked on [lines 239 & 240](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/3518/files/fe889232aa97af99528452eebaf177b9919e5c00#diff-c8dda3f44ea07cbcb7f349be0db86ec4812b54eeae5adc40046c73151d09082aR239-R240). The Component is then cast into a `GameObject` here. I'm wondering if it makes more sense to simply build a list of the `GameObject`s rather than essentially doing the cast twice.